source/opt/value_number_table.cpp

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "value_number_table.h"
+
+#include <algorithm>
+
+namespace spvtools {
+namespace opt {
+
+uint32_t ValueNumberTable::GetValueNumber(spvtools::ir::Instruction* inst) {
+  // TODO: Need to implement the substitution of operands by their value number
+  // before hashing.
+  // TODO: Implement a normal form for opcodes that commute like integer
+  // addition.  This will let us know that a+b is the same value as b+a.
+  assert(inst->result_id() != 0 &&
+         "inst must have a result id to get a value number.");
+
+  // Check if this instruction already has a value.
+  auto id_to_val = id_to_value_.find(inst->result_id());
+  if (id_to_val != id_to_value_.end()) {
+    return id_to_val->second;
+  }
+
+  // If it is a load from memory that can be modified, we have to assume the
+  // memory has been modified, so we give it a new value number.
+  if (inst->IsLoad() && !inst->IsReadOnlyLoad(context())) {
+    uint32_t value_number = TakeNextValueNumber();
+    id_to_value_[inst->result_id()] = value_number;
+    return value_number;
+  }
+
+  // Otherwise, we check if this value has been computed before.
+  auto value = instruction_to_value_.find(*inst);
+  if (value != instruction_to_value_.end()) {
+    uint32_t value_number = id_to_value_[value->first.result_id()];
+    id_to_value_[inst->result_id()] = value_number;
+    return value_number;
+  }
+
+  // If not, assign it a new value number.
+  uint32_t value_number = TakeNextValueNumber();
+  id_to_value_[inst->result_id()] = value_number;
+  instruction_to_value_[*inst] = value_number;
+  return value_number;
+}
+
+bool ComputeSameValue::operator()(const ir::Instruction& lhs,
+                                  const ir::Instruction& rhs) const {
+  if (lhs.opcode() != rhs.opcode()) {
+    return false;
+  }
+
+  if (lhs.type_id() != rhs.type_id()) {
+    return false;
+  }
+
+  if (lhs.NumInOperands() != rhs.NumInOperands()) {
+    return false;
+  }
+
+  for (uint32_t i = 0; i < lhs.NumInOperands(); ++i) {
+    if (lhs.GetInOperand(i) != rhs.GetInOperand(i)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+std::size_t ValueTableHash::operator()(
+    const spvtools::ir::Instruction& inst) const {
+  std::size_t hash = 0;
+  std::hash<uint32_t> hash_function;
+
+  // We hash the in operands only because we are want instructions that are
```

Comment:
s/we are want/we want/.

I'm not following, if you don't hash the opcode, won't you get the same hash value for '%5 * %6' and '%5 + %6'?

Reply:
Yes, I should have included the opcode.